src/util/psbt/map/global.rs

```diff
@@ -113,6 +155,60 @@ impl Map for Global {
             });
         }
 
+        // BIP 174: The Combiner must remove any duplicate key-value pairs, in accordance with
+        //          the specification. It can pick arbitrarily when conflicts occur.
+
+        // Keeping the highest version
+        self.version = cmp::max(self.version, other.version);
+
+        // Merging xpubs
+        for (xpub, (fingerprint1, derivation1)) in other.xpub {
+            match self.xpub.entry(xpub) {
+                Entry::Vacant(entry) => {
+                    entry.insert((fingerprint1, derivation1));
+                },
+                Entry::Occupied(mut entry) => {
+                    // Here in case of the conflict we select the version with algorithm:
+                    // 1) if fingerprints are equal but derivations are not, we report an error; otherwise
+                    // 2) choose longest unhardened derivation; if equal
+                    // 3) choose longest derivation; if equal
+                    // 4) choose the largest derivation using binary ordering; if equal
+                    // 5) check that fingerprints are equal, otherwise fail;
+                    // 6) if fingerprints are also equal we do nothing
+                    let (fingerprint2, len2, normal_len2, deriv_cmp) = {
+                        // weird scope needed for rustc 1.29 borrow checker
+                        let (fp, deriv) = entry.get().clone();
+                        (
+                            fp,
+                            deriv.len(),
+                            deriv.into_iter().rposition(ChildNumber::is_normal),
+                            derivation1.cmp(&deriv)
+                        )
+                    };
+                    let len1 = derivation1.len();
+                    let normal_len1 = derivation1.into_iter().rposition(ChildNumber::is_normal);
+                    match (normal_len1.cmp(&normal_len2), len1.cmp(&len2), deriv_cmp, fingerprint1.cmp(&fingerprint2)) {
+                        (Ordering::Equal, Ordering::Equal, Ordering::Equal, Ordering::Equal) => {},
+                        (Ordering::Equal, Ordering::Equal, Ordering::Equal, _) => {
+                            return Err(psbt::Error::MergeConflict(format!(
+                                "global xpub {} has inconsistent key sources", xpub
+                            ).to_owned()));
+                        }
+                        (Ordering::Greater, ..)
+                        | (Ordering::Equal, Ordering::Greater, ..)
+                        | (Ordering::Equal, Ordering::Equal, Ordering::Greater, _) => {
```

Comment:
I still have the feeling there's  an "inconsistent key source" hiding in here. Does it ever make sense that the same key can be derived from two different parents using equal-length derivation paths? That would make the key-ancestry a DAG instead of a tree, which seems odd (haven't checked the crypto, maybe it could be constructed, but I doubt it).

I guess specifying what we actually want to achieve instead of how we get there would be helpful. From what I gather it would be along the lines of "always choose the most senior key in the ancestry-chain, fail otherwise". 

That would mean checking:
* if fingerprint and derivation path are equal do nothing
* else if one (naturally the shorter) derivation path is a suffix of the other choose the entry of the longer derivation path
* else return an error, something is fishy

Sorry for coming up with this so late, I hope it makes sense. If not I'm not too opposed to potentially not detecting all key source inconsistencies, seems like any serious interactive protocol needs its own much stricter validation before merging potentially attacker-generated PSBTs. But we should be clear on that.

Reply:
No, if you had any "merges" in the key tree that would imply a hash collision.

The only way to get differing paths is for one path to be a truncation of the other (with a prefix collapsed into the fingerprint), which would be visible in the length.

Reply:
Sorry, have not understood do I need to change something here or not. If not, @sgeisler can you pls re-review the PR such that it does not require my changes?

Reply:
This needs changes: `(Ordering::Equal, Ordering::Equal, Ordering::Greater, _)` means that two different derivation paths of the same length exist for the same key, which can't be, so this should probably be an error case.

My broader observation is that this logic is unnecessarily convoluted, making it prone to such mistakes, which is why I propose the simpler algorithm outlined above. I think a simpler algorithm with essentially the same, but now clearly stated goal, would be preferable here but I won't block the PR because of that.

Reply:
Rewrote algorithm taking you idea as a baseline and simplifying it a bit further.

Also, I think that we have to choose longest **hardened** path, since it contain more of valuable information

Reply:
I don't fully understand what you mean with "contain more of valuable information".

If two paths lead to the same key one needs to be a suffix of the other. The only question is which of the both to choose: the longer or the shorter one. Under which circumstance would you choose the shorter one and under which the longer? I guess you mean to make the choice somehow based on hardened derivations in the prefix (part not contained in short path) of the longer path? If so what's the motivation?

Reply:
If we have two paths like `[ac04]/48'/1'/1'/3` and `[8b9a]/1'/3` the first version presents more information since it is not just longer, but also contains more hardened keys in it.

Writing the example I think I understood what you mean: that if for two derivation paths of different lengths the shortest one is not a strict prefix of the other, something is wrong. I see... Let me modify the logic, it will simplify it even more.

Reply:
Yeah, I think given this fact it's probably a sane (not always the best, but you can't please everyone and need to standardize _something_) choice to always take the longer one.

Reply:
@sgeisler pushed reworked version